src/softmax.jl

```diff
@@ -33,20 +33,33 @@ julia> softmax([1, 2, 3])
 See also [`logsoftmax`](@ref).
 """
 softmax(x; dims = 1) = softmax!(similar(x, (float ∘ eltype)(x)), x; dims = dims)
+
 softmax!(x; dims = 1) = softmax!(x, x; dims = dims)
+
 function softmax!(out::O, x::T; dims = 1) where {O<:AbstractArray,T<:AbstractArray}
     out .= exp.(x .- maximum(x; dims = dims))
     out ./= sum(out; dims = dims)
 end
 
-∇softmax(Δ, x; dims = 1) = ∇softmax!(similar(Δ), Δ, x; dims = dims)
-∇softmax!(Δ, x; dims = 1) = ∇softmax!(Δ, Δ, x; dims = dims)
-function ∇softmax!(out::O, Δ::O, x::T; dims = 1) where {O<:AbstractArray,T<:AbstractArray}
-    softmax!(out, x; dims = dims)
-    out .*= Δ .- sum(Δ .* out; dims = dims)
+∇softmax(Δ, x, y; dims = 1) = ∇softmax!(similar(Δ), Δ, x, y; dims = dims)
+
+## Can introduce at the end of deprecation cycle of ∇softmax!(out, Δ, x; dims = 1)  
+#∇softmax!(Δ, x, y; dims = 1) = ∇softmax!(Δ, Δ, x, y; dims = dims)
+
+function ∇softmax!(out::AbstractArray, Δ::AbstractArray, 
+                    x::AbstractArray, y::AbstractArray; dims = 1)
+    out .*= Δ .- sum(Δ .* y; dims = dims)
```

Comment:
Don't you need `out .= y` here too? Since `∇softmax!(similar(Δ), Δ, x, y; dims = dims)` produces garbage. 

I also wonder whether it would be safe to do `sum(y .= Δ .* y; dims = dims)` to save the allocation? By the time we are executing this, everything downstream of `y` ought to be finished.

Reply:
good catch

Reply:
I guess one could use `mapreduce(*, +,   Δ, y; dims)` to avoid `Δ .* y`... although right now this doesn't do anything efficient, sadly. (e.g. `mapreduce(*, +, fill(3, 2,2), fill(5, 2,2), dims=1)`.)